controllers/k8ssandra/replication.go

```diff
@@ -50,24 +78,209 @@ func (r *K8ssandraClusterReconciler) checkSystemReplication(ctx context.Context,
+		err = fmt.Errorf("the %s annotation must include all user keyspaces for each specified DC", api.DcReplicationAnnotation)
+		logger.Error(err, "Invalid "+api.DcReplicationAnnotation+" annotation")
+		return result.Error(err)
+	}
+
+	replication = getReplicationForDeployedDcs(kc, replication)
+
+	logger.Info("computed replication")
+
+	for _, ks := range userKeyspaces {
+		replicationFactor := replication.ReplicationFactor(dc.Name, ks)
+		logger.Info("computed replication factor", "keyspace", ks, "replication_factor", replicationFactor)
+		if replicationFactor == 0 {
+			continue
+		}
+		if err = ensureKeyspaceReplication(mgmtApi, ks, dc.Name, replicationFactor); err != nil {
+			logger.Error(err, "Keyspace replication check failed", "Keyspace", ks)
+			return result.Error(err)
+		}
+	}
+
+	return result.Continue()
+}
+
+func getUserKeyspaces(mgmtApi cassandra.ManagementApiFacade, kc *api.K8ssandraCluster) ([]string, error) {
+	keyspaces, err := mgmtApi.ListKeyspaces("")
+	if err != nil {
+		return nil, err
+	}
+
+	internalKeyspaces := getInternalKeyspaces(kc)
+	userKeyspaces := make([]string, 0)
+
+	for _, ks := range keyspaces {
+		if !utils.SliceContains(internalKeyspaces, ks) {
+			userKeyspaces = append(userKeyspaces, ks)
+		}
+	}
+
+	return userKeyspaces, nil
+}
+
+// getReplicatedInternalKeyspaces returns Cassandra, internal keyspaces that should be
+// replicated, namely system_auth. The slice also includes the Stargate auth keyspace if
+// Stargate is enabled and the Reaper keyspace if Reaper is enabled.
+func getReplicatedInternalKeyspaces(kc *api.K8ssandraCluster) []string {
+	keyspaces := []string{"system_traces", "system_distributed", "system_auth"}
+
+	if kc.HasStargates() {
+		keyspaces = append(keyspaces, stargate.AuthKeyspace)
+	}
+
+	if kc.HasReapers() {
+		if kc.Spec.Reaper != nil && kc.Spec.Reaper.Keyspace != "" {
+			keyspaces = append(keyspaces, kc.Spec.Reaper.Keyspace)
+		} else {
+			keyspaces = append(keyspaces, reaperapi.DefaultKeyspace)
+		}
+	}
+
+	return keyspaces
+}
+
+// getInternalKeyspaces returns all internal Cassandra keyspaces as well as the Stargate
+// auth and Reaper keyspaces if Stargate and Reaper are enabled.
+func getInternalKeyspaces(kc *api.K8ssandraCluster) []string {
+	keyspaces := getReplicatedInternalKeyspaces(kc)
+	keyspaces = append(keyspaces, "system", "system_schema")
+	return keyspaces
+}
+
+// getReplicationForDeployedDcs gets the replication for only those DCs that have already
+// been deployed. The replication argument may include DCs that have not yet been deployed.
+func getReplicationForDeployedDcs(kc *api.K8ssandraCluster, replication *cassandra.Replication) *cassandra.Replication {
+	dcNames := make([]string, 0)
+	for _, dc := range kc.Spec.Cassandra.Datacenters {
+		if status, found := kc.Status.Datacenters[dc.Meta.Name]; found && status.Cassandra.GetConditionStatus(cassdcapi.DatacenterInitialized) == corev1.ConditionTrue {
+			dcNames = append(dcNames, dc.Meta.Name)
+		}
+	}
```

Comment:
```suggestion
	for _, template := range kc.GetReadyDatacenters() {
		dcNames = append(dcNames, template.Meta.Name)
	}
```

Reply:
Once we add support for stopping a DC (#296) it would be wrong to check only the Ready condition. We will need to check the Initialized condition. How about I add a `GetInitializedDatacenters` method?

Reply:
Ah right. Yes, that works for me.